### PR TITLE
[node-core-library] Include functionality to handle processes terminating via a signal.

### PR DESCRIPTION
**Pull request by iclanton**
## Summary

- Add a `throwOnSignal` option to the `Executable.waitForExitAsync` to control if that function should throw if the process is terminated with a signal
- Add a `signal` property to the result of `Executable.waitForExitAsync` that includes a signal if the process was termianted by a signal.

## Details

This returns the function to the behavior before https://github.com/microsoft/rushstack/pull/4721. The current (post-https://github.com/microsoft/rushstack/pull/4721) behavior can be retained by including the `throwOnSignal: true` option. Note that all callers of `waitForExitAsync` maintain the current behavior.

## How it was tested

Included a unit test.

## Impacted documentation

None.

### CODE REVIEW

File: libraries/node-core-library/src/test/test-data/executable/no-terminate/javascript-file.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const [node, script, ...args] = process.argv;
+console.log(`Executing no-terminate with args: ${args.join(' ')}`);
+
+console.error('This process never terminates');
+
+const { copySync } = require('fs-extra');
```
Comment (dmichon-msft) [inline libraries/node-core-library/src/test/test-data/executable/no-terminate/javascript-file.js:6]: Unused import?

**Review by dmichon-msft — APPROVED** [review]
Just a rogue unused import in the test script

File: libraries/node-core-library/src/test/Executable.test.ts [hunk 2]
```diff
@@ -298,17 +284,44 @@ describe('Executable process tests', () => {
...
+    const childProcess: child_process.ChildProcess = Executable.spawn(
+      process.argv0,
+      [executablePath, '1', '2', '3'],
+      {
+        environment,
+        currentWorkingDirectory: executableFolder
+      }
+    );
+    const startTime: number = Date.now();
+    await Async.sleepAsync(50); // Wait for the process to boot and print
```
Comment (D4N14L) [inline libraries/node-core-library/src/test/Executable.test.ts:308]: Should we wait for the line to be printed by listening to the stdout?

File: libraries/node-core-library/src/test/Executable.test.ts [hunk 3]
```diff
@@ -298,17 +284,44 @@ describe('Executable process tests', () => {
...
+    });
+    const endTime: number = Date.now();
+    expect(killResult).toBe(true);
+    expect(result.signal).toBe('SIGTERM');
+    expect(result.exitCode).toBeNull();
+    expect(typeof result.stdout).toEqual('string');
+    expect(typeof result.stderr).toEqual('string');
+    expect(result.stdout).toMatch(/^Executing no-terminate with args:/);
+    expect(result.stderr.includes('This process never terminates')).toBe(true);
+    expect(endTime - startTime).toBeLessThan(500);
```
Comment (D4N14L) [inline libraries/node-core-library/src/test/Executable.test.ts:321]: This feels like something that could possibly become flaky on different hardware

**Review by D4N14L — APPROVED** [review]

File: libraries/node-core-library/src/test/Executable.test.ts (hunk 2 above)
Comment (iclanton) [reply libraries/node-core-library/src/test/Executable.test.ts:308]: If it becomes a problem, we'll change this.

File: libraries/node-core-library/src/test/Executable.test.ts (hunk 3 above)
Comment (iclanton) [reply libraries/node-core-library/src/test/Executable.test.ts:321]: If it becomes a problem, we'll change this.

**Review by D4N14L — APPROVED** [review]

**Pull request merged**